### PR TITLE
test: prototype faster test runner

### PR DESCRIPTION
**Pull request by bartlomieju**

### CODE REVIEW

**Comment by bartlomieju** [comment]
@ry before this PR std test were run only with release build, now CI fails on std tests with debug build... Can this `std_tests` test be run only in release tests?

**Comment by ry** [comment]
Yeah we can conditionally skip the std test if it’s a debug build. Please add a TODO to fix it.

File: std/testing/runner.ts [hunk 1]
```diff
@@ -168,6 +191,19 @@ export async function runTestModules({
     return;
   }
 
+  const testFile = renderTestFile(testModules);
+  const root = Deno.env("DENO_DIR") || Deno.cwd();
+  const testFilePath = join(root, ".deno.test.ts");
```
Comment (nayeemrmn) [inline std/testing/runner.ts]: `Deno.makeTempDir()`? Avoids unnecessary env usage and user space code shouldn't use `DENO_DIR` like this IMO.

Comment (bartlomieju) [reply std/testing/runner.ts]: Actually this is intentional - I should have commented it. 

This actually tricks Deno to not recompile files if there's no need. Let me explain:
1. When you first run `DENO_DIR/.deno.test.ts` Deno will compile it and cache it
2. The file is removed by test runner
3. On next test run file is created again - if you did not add new files that runner picks up and files did not change compiler will recompile them
4. If any of test files changes Deno will pick it up during loading the module into V8 and recompile only this single module

Comment (nayeemrmn) [reply std/testing/runner.ts]: Oh of course, the cache would be cluttered with temp modules otherwise.

File: cli/tests/std_tests.rs [hunk 2]
```diff
@@ -1,25 +1,34 @@
...
-    .arg("std")
-    .spawn()
-    .expect("failed to spawn script");
-  let status = deno.wait().expect("failed to wait for the child process");
-  assert_eq!(Some(0), status.code());
-  assert!(status.success());
+  // TODO: this is hacky and should conditionally skipped
+  // some more idiomatic way
+  // Run only with release build
+  if let Ok(mode) = env::var("DENO_BUILD_MODE") {
```
Comment (ry) [inline cli/tests/std_tests.rs]: Use `#[cfg(debug_assertions)]`

File: cli/tests/std_tests.rs [hunk 3]
```diff
@@ -1,34 +1,32 @@
...
-fn std_tests() {
-  // TODO: this is hacky and should conditionally skipped
-  // some more idiomatic way
-  // Run only with release build
-  if let Ok(mode) = env::var("DENO_BUILD_MODE") {
-    if mode == "release" {
-      let dir = TempDir::new().expect("tempdir fail");
-      let mut deno_cmd = Command::new(deno_exe_path());
-      deno_cmd.env("DENO_DIR", dir.path());
+// Runs only on release build
```
Comment (ry) [inline cli/tests/std_tests.rs]: Please add `// TODO fix tests in release mode.`

File: cli/tests/std_tests.rs [hunk 4]
```diff
@@ -1,34 +1,32 @@
...
+  #[test]
+  fn std_tests() {
+    let dir = TempDir::new().expect("tempdir fail");
+    let mut deno_cmd = Command::new(deno_exe_path());
+    deno_cmd.env("DENO_DIR", dir.path());
+
+    let mut cwd = root_path();
+    cwd.push("std");
+    let mut deno = deno_cmd
+      .current_dir(cwd) // TODO: std tests expect to run from "std" dir
```
Comment (ry) [inline cli/tests/std_tests.rs]: s/TODO/note/

File: cli/tests/std_tests.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/tests/std_tests.rs]: In debug mode? Working in release

File: std/testing/runner.ts (hunk 1 above)
Comment (bartlomieju) [reply std/testing/runner.ts]: Added a few paragraphs about what's happening there.

File: cli/tests/std_tests.rs (hunk 3 above)
Comment (ry) [reply cli/tests/std_tests.rs]: er, yes that's what I meant.

**Comment by bartlomieju** [comment]
Oh this is soooo cool
```
DEBUG RS - deno::isolate:294 - dyn_import specifier C:\Users\RUNNER~1\AppData\Local\Temp\.tmpGMWvcR\.deno.test.ts referrer file:///D:/a/deno/deno/std/testing/runner.ts 
DEBUG RS - deno_cli::state:310 - state permissions denied for dyn import "c:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\.tmpGMWvcR\\.deno.test.ts"
DEBUG RS - deno::isolate:503 - dyn_import_done 0 Err(Some("permission denied"))
⚠️️  Granted write access to "C:\Users\RUNNER~1\AppData\Local\Temp\.tmpGMWvcR\.deno.test.ts"
```

`std_tests` are run with `-A` flag and even though dynamic permission is denied for just created file. Interesting...

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -118,18 +118,9 @@ jobs:
...
 
-      # TODO(ry) Remove this step, and move the following test to
-      # cli/tests/std_tests.rs
-      # TODO(ry) Remove the "cd std".
-      - name: std test
-        if: matrix.kind == 'test_std'
-        run: |
-          cd std
-          ../target/release/deno test -A
-
```
Comment (ry) [inline .github/workflows/ci.yml]: \o/ thank you

File: std/testing/runner.ts [hunk 6]
```diff
@@ -116,6 +116,28 @@ export interface RunTestModulesOptions extends RunTestsOptions {
...
+
+  for (const testModule of testModules) {
+    // NOTE: this is intentional that template string is not used
+    // because of TS compiler quirkness of trying to import it
+    // rather than treating it like a variable
+    testFile += 'import "' + testModule + '"\n';
+  }
+
+  return testFile;
+}
```
Comment (ry) [inline std/testing/runner.ts]: This is a hack but let's just go with it for now. It's much better to have the std tests running in "cargo test"

**Pull request merged**